### ISSUE TITLE
Getting exports into core

### ISSUE THREAD

**Issue by jleclanche** [translate/pootle#3532]
Was talking with @julen on IRC earlier and he mentioned Evernote may want to move their internal TTX exports to Pootle. This overlaps with work I'm currently doing to bring zip exports back into master.

The zip export use cases are a stand-in for offline translations and what we need to support on our side is:
 * Exporting arbitrary pootle paths (store, directory, language, project)
 * Importing those resources back into pootle

Points of note:
* Separation of concerns. Imports and exports should be separate, probably.
* Ability to disable exports/imports
* Permissions. For exports, probably not. For imports, probably so.
* Pluggable backends (export to arbitrary formats)
* TTX exports: are they relevant for directory/language/project (multi-file)?
 * Is what we're even doing "zip exports"? Isn't it "zipped po exports"? We don't use zip for single-file, do we?
 * Or maybe call it "original format export"? God I hate that we don't have format-agnostic stores
* UX for RQ feedback when the user requests an export (takes time, how do we notify them? etc)

**Comment by jleclanche** [comment]
cc @iafan @dwaynebailey

**Comment by unho** [comment]
About the UX for exports, my idea is to do it the Launchpad way, which is providing a link to "request the download" and not a link for "download". When the file is ready to download a link is emailed to the requesting user. Since generating this stuff is expensive, and probably it is mostly done for offline translation (and not as backup) I assume we can safely allow this only to logged in users. Of course we will have to take in account users not having email (it might be a good idea to enforce this to be always set in order to build communities). Once this is in place we can change the link to "download" if the file is already generated.

**Comment by dwaynebailey** [comment]
Here's my feedback:

General comments
-------------------------

Lets not conflate whatever needs there are for TTX export with this. I see massive scope creep when the objective for Pootle is to do the following:

1. Recover the lost functionality
2. Improve the performance and bottlenecks associated with this.

Note we at the moment only every exorted store and directory.  I'm almost certain we have never exported language or project.  These later might make sense in e.g. doing whole lang offline, or wanting to pull all translations for a manual commit to VCS for a project.  But be aware that's a step beyond just recovery.

Separation of concerns: yes import and export are different. But we can't separate the fact that we need both.  Export is simply sync and zip.  Import is unzip and merge.  The merge part is harder but exists.

Disable: yes that would be a given since this is also about modularisation/plugins for any of the bits that we recover.

Permissions: exports I'd say yes, but not a new one.  I say yes only because we don't want arbitrary load from not logged in users.  For import nothing special, if you can translate or suggest you should be able to import.  The only special perm we had was the ability to overwrite which we might want to bring back.  In an ideal world we'd do merging much better i.e we'd know the state of the unit when we downloaded so that we can determine if target changed while we were offline in which case a suggestion otherwise a valid translation.

Pluggable backends for arbitrary format.  I think that's best dealt with when we arrive at format agnostic.  The only real value is PO, XLIFF and TTX where we're exporting to match what the user might have for offline translation.  Other formats would be counterproductive in terms of offline translation.  Makes sense if someone wanted .properties for testing though but then we're dealing with whole project download and monolingual files in a format agnostic world I think.

TTX export: not in scope I think.  Its a zip of the underlying format.  So I think if the format is XLIFF you'd get a zip of XLIFF files.  We don't zip individual files no.  But for a single file we still need to export from the DB into the format and give that to the user, if the file is big (probably really big) that potentially takes a long time.  Yip so "original format export" is the right wording. Ironically we kinda are format agnotic in that the DB just has the strings that we serialise to the defined format.

UX ideas
------------

- I'm dead against emailing.  I think launchpad does that for serious load reasons. We have load but not that level of load.  Taking people out of Pootle to email is I think the wrong approach.

- I'd split this into two stage:
  
  1. Do what we do now
  2. Add RQ to the mix

- With RQ in the mix

  1. User clicks on link
  2. Job sent to RQ
  3. They get the zip/file if it take <1s
  4. An orange message appears at the top like for refresh stats saying something like "Preparing your files"
  5. Message changes to green "Files ready or download"
  6.  User clicks on that link to get the files

  - The user can click multiple times on the link.  We only load a new job if one doesn't exist already.
  - If a different user requests the exact same file.  We simply link them to the same queue item.

**Comment by julen** [comment]
I believe too many details are mixed here, because the shape of what you want to export is job-specific, and what you want to shape up is the overall process of how an exporting workflow works.

**Comment by jleclanche** [comment]
> Permissions: exports I'd say yes, but not a new one. I say yes only because we don't want arbitrary load from not logged in users.

We can restrict to logged-in users without permsisions.
